vllm/triton_utils/jit_cache.py

```diff
@@ -0,0 +1,417 @@
+        self.grid_obj = grid_obj
+        self.grid_is_callable = callable(grid_obj)
+        self.grid_size = len(
+            grid_example
+        )  # grid_example is always not callable, so we need both
+        self.cache_launch_grid = cache_launch_grid
+        self.concrete_grid = (None, None, None)
+        if cache_launch_grid:
+            grid_0 = grid_example[0]
+            grid_1 = grid_example[1] if self.grid_size > 1 else 1
+            grid_2 = grid_example[2] if self.grid_size > 2 else 1
+            self.concrete_grid = (grid_0, grid_1, grid_2)
+        self.kernel = kernel
+        self.launch_metadata = launch_metadata
+        self.launch_enter_hook = launch_enter_hook
+        self.launch_exit_hook = launch_exit_hook
+
+        self.arg_list = []
+        self.update_args_index = {}
+        # We construct the list of arguments that are passed to the combiled
+        # kernel beforehand. For the arguments that could change each time the
+        # kernel is called, store a dummy value that will be set each time
+        # __call__ is called. For the arguments that are labelled as assume to
+        # be constant, we skip this step and use the initial stored values.
+        for i, arg_n in enumerate(bound_args.keys()):
+            if arg_n in update_only_arg_names:
+                self.update_args_index[arg_n] = i
+                self.arg_list.append("dummy_value")
+            else:
+                self.arg_list.append(bound_args[arg_n])
+
+        self.device = device
+        self._init_handles()
+        self.cache_key = cache_key
+
+    def _init_handles(self):
+        """
+        more or less redo what CompiledKernel._init_hanles is doing
+        (c.f. triton/python/triton/runtime/compiler.py:379)
+        """
+        self.run = driver.active.launcher_cls(self.kernel.src,
+                                              self.kernel.metadata)
+        # check once and not again
+        self.dev_max_shared = driver.active.utils.get_device_properties(
+            self.device)["max_shared_mem"]
+        if self.kernel.metadata.shared > self.dev_max_shared:
+            raise OutOfResources(self.kernel.metadata.shared,
+                                 self.dev_max_shared, "shared memory")
+        self.module, self.function, self.n_regs, self.n_spills = (
+            driver.active.utils.load_binary(
+                self.kernel.name,
+                self.kernel.kernel,
+                self.kernel.metadata.shared,
+                self.device,
+            ))
+
+    def __call__(self, *args, **kwargs):
+        assert len(args) == 0
+
+        for arg_n, idx in self.update_args_index.items():
+            self.arg_list[idx] = kwargs[arg_n]
+
+        if self.cache_launch_grid:
+            grid_0, grid_1, grid_2 = self.concrete_grid
+        else:
+            if self.grid_is_callable:
+                grid = kwargs["grid"](kwargs)
+            else:
+                grid = copy.deepcopy(kwargs["grid"])
```

Comment:
Is there a downside to always using `grid = copy.deepcopy(kwargs["grid"])`? It would be better to simplify the code if this isn't a significant optimization

Reply:
The first option calls `grid` so it's not the same code